### ISSUE TITLE
e2e-test-utils does not help much with creating admin URLs

### ISSUE THREAD

**Issue by tomalec** [WordPress/gutenberg#34044]
## What problem does this address?

I'd like to make use of e2e utils to go to the admin page of my plugin, precisely `http://localhost:8084/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsetup-mc'`, out of my plugin-specific path (`/google/setup-mc/`)

I tried to use `utils` to make that easier, but they happen to give more confusion and frustration than value, as in the end I need my own helper to construct full URL from the relative admin page path.

I [found there](@wordpress/e2e-test-utils) are three utils/[variables](https://www.npmjs.com/package/@woocommerce/e2e-utils) that could potentially help me:
- > `visitAdminPage()` Visits admin page; if user is not logged in then it logging in it first, then visits admin page.
- > `createURL()` Creates new URL by parsing base URL, WPPath and query string.
- > `WP_ADMIN_WC_HOME` WooCommerce home screen

### TL;DR
None of them was useful, I guess both functions need fixes/improvements. I ended up creating my own util. See the proposed solution at the bottom.

### Chapter 1 - `visitAdminPage`
I started with `visitAdminPage` as from the name & description it seemed to be exactly what I want: get a relative path to the admin page, and go there (log in if needed)

```js
// first guess
visitAdminPage( undefined, { path: '/google/setup-mc' } ) 
// Nope. RTM, query should be stringified already, and `adminPath` is required
```
Ok, I found and read the JSDoc. I need to get to `admin.php?page=wc-admin`, so I bet the "admin page" mentioned in the docs is `wc-admin` .
```js
visitAdminPage( 'wp-admin', '?path=/google/setup-mc' ) 
// Nope. `wp-admin/wc-admin?path=/google/setup-mc` is not what I wanted
```

Uh,, I don't want to hard-code `admin.php?page=wc-admin` everywhere, so I found [`@woocommerce/e2e-utils`](https://www.npmjs.com/package/@woocommerce/e2e-utils) exposes `WP_ADMIN_WC_HOME = 'http://localhost:8084/wp-admin/admin.php?page=wc-admin'` Great! Let's use it.

```js
visitAdminPage( WP_ADMIN_WC_HOME, '?path=/google/setup-mc' ) 
// Nope. 'http://localhost:8084/wp-admin/http:/localhost:8084/wp-admin/admin.php%3Fpage=wc-admin?path=/google/setup-mc' 
```
**`http://localhost:8084/wp-admin/http:/localhost:8084/wp-admin/admin.php%3Fpage=wc-admin?path=/google/setup-mc`**, Seriously?! what kind of admin page is that?

Ok, I give up with `visitAdminPage`, I can leave without logging if needed, it seems I need to create a URL myself

### Chapter 2 - `createURL`

We have lower-level util, just to create URLs, not admin, no login, no `page.goto`, just URL construction. Takes path, takes query string, gives you full URL, based on WP URL - Cool, I can work with that.

```js
// first try
createURL( WP_ADMIN_WC_HOME, '?path=/google/setup-mc' ) 
// Nope. 'http://localhost:8084/http:/localhost:8084/wp-admin/admin.php%3Fpage=wc-admin?path=/google/setup-mc'
```
argh.	.. only relative paths again.

Ok, I'm running out of time, let's just hardcode this WC path.

```js
// first try
createURL( /wp-admin/admin.php?page=wc-admin, '?path=/google/setup-mc' ) 
// Nope. query params are messed up.
```

### Chapter 3 - create it myself from `WP_ADMIN_WC_HOME`

Now, I stepped back, and thought: I got the `WP_ADMIN_WC_HOME` which is almost there, I just need to add more query params there, like:

```js
const adminURL = new URL( WP_ADMIN_WC_HOME );
adminURL.searchParams.set( 'path', '/google/setup-mc');
// 'http://localhost:8084/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsetup-mc'
```

🎉 And it even does escape `/` for me!

Now, my question is why `createURL` and `visitAdminPage` doesn't do exactly that?!
But instead does all those `path.join`, `.search = `, `.pathname = ` which are less flexible, prone to errors, etc.?

## What is your proposed solution?

I'd like to propose something to mitigate the frustrations like mentioned above, as I believe I'm not the only one here, struggling to use those helpers. The ideas in my mind are:
1. Allow, relative as well as absolute paths in both functions
 -  If there are any reasons, why we really need them to accept only relative ones and produce those `…/wpadmin/https://localhost…`  results:
	   - Please mention them in the docs
	   - at least expose constants that are relative, so we could use it like `visitAdminPage( WP_ADMIN_WC_HOME )`
2. Make path an optional parameter
3. Improve developer ergonomics by not requiring the consumer of those functions, to resolve whether the path has search params or not already, so one could do
	```js
	const path = "/foo?bar=baz";
	// …
	const createURL( path, 'fiz=biz' );
	```
	without the need to worry about `?` vs `&`.
4. (nice to have) improve DevX by accepting an object of search params, like `createURL( WP_ADMIN_WC_HOME, { path: '/google/setup-mc', foo: 'bar' } );`

Without the above, IMHO, those helpers does not bring more value than
- `WP_ADMIN_WC_HOME + '&' + query`
- `WP_BASE_URL` + path + '?' + query

#### Additional notes:
1-2 are addressed by https://github.com/WordPress/gutenberg/pull/34045
3-4 by https://github.com/WordPress/gutenberg/pull/34051

**Comment by swissspidy** [comment]
Since you want to access `wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsetup-mc`, have you tried `visitAdminPage( 'admin.php', '?path=/google/setup-mc' )`? That's how the function is intended to be used.

**Comment by tomalec** [comment]
> Since you want to access `wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsetup-mc`, have you tried `visitAdminPage( 'admin.php', '?path=/google/setup-mc' )`? That's how the function is intended to be used.

I did and it doesn't work. (So it seems that API is confusing to you as well ;) )

1. What you suggested produces `http://localhost:8084/wp-admin/admin.php?path=/google/setup-mc'` not `http://localhost:8084/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsetup-mc`
2. I don't want to hard-code `admin.php?page=wc-admin`, as it is not set in my repo/plugin, and would like to get that path from utils, like `WP_ADMIN_WC_HOME`. 
3. Moreover, with the current shape of those helpers `visitAdminPage( 'admin.php?page=wc-admin', '?path=/google/setup-mc' )` will also not work, as it will not join query params and produce `http://localhost:8084/wp-admin/admin.php%3Fpage=wc-admin?path=/google/setup-mc`

	  So, I'd rather have to use:
	  ```js
	  visitAdminPage( 'admin.php', '?page=wc-admin&'+ query )
	  ```
	  What to me looks, like:
	  ```js
	  visitAdminPage( hardcoded_foreign_value_part_1, hardcoded_foreign_value_part_2 + my_query )
	  ```

Please see https://github.com/WordPress/gutenberg/pull/34045 and https://github.com/WordPress/gutenberg/pull/34051. If we merge those changes, the above case could be easily solved like:
```js
// As a WooCommerce developer, setting WP admin pages explicitely
visitAdminPage( 'admin.php', { page: wc-admin, path: '/google/setup-mc' } );
// As a WooCommerce plugin developer, without hard-coding any foreign constants.
visitAdminPage( WP_ADMIN_WC_HOME, { path: '/google/setup-mc' } );
page.goto( createURL( WP_ADMIN_WC_HOME, { path: '/google/setup-mc' } ) );
```

**Comment by swissspidy** [comment]
> I did and it doesn't work. (So it seems that API is confusing to you as well ;) )
> 
> 1. What you suggested produces `http://localhost:8084/wp-admin/admin.php?path=/google/setup-mc'` not `http://localhost:8084/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsetup-mc`

Okay, I forgot to add `page=wc-admin` in my example and to wrap things in `encodeURIComponent`, but otherwise this should just work.

The main thing I wanted to highlight is the `admin.php` part as you did not try this at all.

> Please see #34045 and #34051. If we merge those changes, the above case could be easily solved like:

Supporting something like `visitAdminPage( 'admin.php', { page: wc-admin, path: '/google/setup-mc' } );` would be a nice enhancement 👍

**Comment by tomalec** [comment]
> Okay, I forgot to add `page=wc-admin` in my example and to wrap things in `encodeURIComponent`, but otherwise this should just work.

IMHO, having to cater for:
 - Whether the path is relative or absolute
 - Resolving manually absolute path to relative to the **unexposed** `WP_BASE` (so `createURL` could then resolve it aginst from relative to absolute based on `WP_BASE`)
 - `encodeURIComponent`
 - concatenating query params

Does not leave much of a value in `createURL`. Other than exposing `WP_BASE`, but for that, I'd rather use it like:
```js
import { WP_BASE_URL } from '@wordpress/e2e-utils'

const url = new URL( WP_ADMIN_WC_HOME, WP_BASE_URL );
url.searchParams.set( 'path', '/google/setup-mc' );
```
...than what I have to [do currently](https://github.com/woocommerce/google-listings-and-ads/blob/add/886-e2e-jetpack/tests/e2e/utils/create-url.js#L10-L11)
```js
import { createURL } from '@wordpress/e2e-utils';
const WP_BASE_URL = createURL( '' );

const url = new URL( WP_ADMIN_WC_HOME, WP_BASE_URL );
url.searchParams.set( 'path', '/google/setup-mc' );
```

**Comment by swissspidy** [comment]
Not disagreeing on the current state of `createURL` and your proposed enhancement. Just saying that `visitAdminPage` is pretty straightforward to use and works well. You just need to pass `admin.php` or similar instead of a path or a full URL.

**Comment by kevin940726** [comment]
Hi! Thanks for the thoughtful issue. I agree with you that with the help of the modern native API, some utils might not be needed after all. And that's totally OK, you can just use `new URL()` and ignore the utils in your tests. We're in the process of migrating from Puppeteer to Playwright, and for historical reasons, we still have some legacy utils like [`visitAdminPage`](https://github.com/WordPress/gutenberg/blob/2929720235e4f8537424b0a4fcbde4a06eda23a2/packages/e2e-test-utils-playwright/src/admin/visit-admin-page.ts#L18). We plan to ultimately remove the need of them, and just accepts URL in `goto`. Playwright's API also makes it easier as it doesn't require specifying the base url every time. Note that the package hasn't been released yet though.
